types/lodash/common/function.d.ts

```diff
@@ -1372,25 +1375,34 @@ declare module "../index" {
          * @param options.trailing Specify invoking on the trailing edge of the timeout.
          * @return Returns the new throttled function.
          */
-        throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettings): DebouncedFunc<T>;
+        throttle<T extends (...args: any) => any>(func: T, wait: number | undefined, options: ThrottleSettingsNoLeading): DebouncedFunc<T>;
+        throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettings): DebouncedFuncLeading<T>;
```

Comment:
This is a little awkward, since this overload will match if the caller passes a boolean variable (instead of the literal `this` for `leading`. In this case, we don't know if it's leading or not, so ideally the return type should include undefined just in case.

Is there a way to check if the `leading` field is unset? e.g.

```ts
ThrottleSettingsLeadingUnset extends ThrottleSettings {
  leading?: never; // or maybe undefined?
}
```

Then return `DebouncedFuncLeading` only if `leading` is unset or set to true.

Reply:
Good point! As an extra complication, I just checked and apparently lodash treats an explicit `{ leading: undefined }` the same as `{ leading: false }`. So unset and undefined fields should be treated differently.

This is what I came up with, going to commit with tests but not sure if there's a way to handle this without some kind of union type. `never` didn't seem to work.

```ts
interface ThrottleSettings {
    leading?: boolean | undefined;
    trailing?: boolean | undefined;
}
type ThrottleSettingsLeading = (ThrottleSettings & { leading: true }) | Omit<ThrottleSettings, 'leading'>
interface LoDashStatic {
    throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettingsLeading): DebouncedFuncLeading<T>;
    throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettings): DebouncedFunc<T>;
}
```

Reply:
Wow good find! In that case it might be nice to discourage people from setting the field to `undefined` since it almost certainly doesn't do what they expect. But I don't think there's a good way to do that (at least not better than what you have already) so I think we'll have to settle for this.

Reply:
I just pushed a commit to clarify the tests. Had the thought right after that it _is_ possible to prohibit an explicit undefined here, while still making `leading` an optional property.

```ts
interface ThrottleSettings {
    leading: boolean;
    trailing?: boolean | undefined;
}
type ThrottleSettingsLeading = (ThrottleSettings & { leading: true }) | Omit<ThrottleSettings, 'leading'>
interface LoDashStatic {
    throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettingsLeading): DebouncedFuncLeading<T>;
    throttle<T extends (...args: any) => any>(func: T, wait?: number, options?: ThrottleSettings): DebouncedFunc<T>;
}
```

But I think that's a breaking change and I don't know if it's preferable anyway.

Reply:
Pretty much every type change is a breaking change, due to the way types work. But I think that change still doesn't completely prevent explicit undefined, so it's debatable whether it's worth the potential breakage. (It prevents explicit undefined in object literals but not prebuilt objects. Maybe better than nothing?)